core/aws-core/src/main/java/software/amazon/awssdk/awscore/client/builder/AwsDefaultClientBuilder.java

```diff
@@ -422,6 +424,20 @@ private void configureRetryPolicy(SdkClientConfiguration.Builder config) {
     private void configureRetryStrategy(SdkClientConfiguration.Builder config) {
         RetryStrategy strategy = config.option(SdkClientOption.RETRY_STRATEGY);
         if (strategy != null) {
+            // Fix the retry strategy if no retry predicates were configured. Users using
+            // DefaultRetryStrategy do not have any conditions configured as that package
+            // does not know anything about the SDK or AWS retry conditions. This mimics the
+            // retry policies behavior that are configured behind the scenes and avoids the
+            // risk of customers inadvertently using a no-op retry strategy.
+            if (strategy.maxAttempts() > 1
+                && (strategy instanceof BaseRetryStrategy)
+                && !((BaseRetryStrategy) strategy).hasRetryPredicates()
+            ) {
+                RetryStrategy.Builder<?, ?> builder = strategy.toBuilder();
+                SdkDefaultRetryStrategy.configureStrategy(builder);
+                AwsRetryStrategy.configureStrategy(builder);
+                config.option(SdkClientOption.RETRY_STRATEGY, builder.build());
+            }
```

Comment:
The condition in this "if" feels fragile. If the thing we're trying to fix is "the customer thinks they are creating the default strategy and adding to it", this wouldn't address it. i.e. them adding predicates means they don't get any of the built-in ones.

Should we have an equivalent of "can the client add its own defaults?" to `RetryStrategy`? We could default it to false in the interface and true in our strategies. We'd use that method in the if statement instead of these conditions. If a customer wants to exclude the client-provided conditions, they can set the flag to false.

Reply:
Not a big issue but we might end up with duplicated conditions. Say, they are using the correct way of building an AWS retry strategy but they just want to increase the number of retries, they would do

```
.overrideConfiguration(o -> o.retryStrategy(AwsRetryStrategy.standardRetryStrategy()
 .toBuilder()
 .maxAttemtps(10)
 .build()))
```

So we will add the default conditions twice. 

Also, the flag cannot be toggled unless is part of the state of the strategy and set in the builder, or they would have to override the method. Is that better than, say, just add the defaults here unconditionally?